FreeBSD 13.2 Build failures

Bitrot. Compiled in January 2023. Passed almost all tests; failed 3, I think, network related tests (due to differing OS
specific device names).

FreeBSD 13.2 with current updates
Java 8
Scala Native 0.5.0-SNAPSHOT
Scala 3.3.1
LLVM-15
GC: commix, immix gives similar results. Have not tried Boehm
command: sbt> sandbox3/run

Something to peck away at whilst waiting for CI runs.  

This Issue may or may not be related to my recent series of `Discover.scala` Issues.

The `MAP_NORESERVE` probably will take the most study. 


```[info] Generating intermediate code (with debug metadata) (9259 ms)
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/gc/shared/Safepoint.c:28:14: error: use of undeclared identifier 'MAP_NORESERVE'
[error]              MAP_NORESERVE | MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
[error]              ^
[error] 1 error generated.
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:60:9: warning: '__noinline' macro redefined [-Wmacro-redefined]
[error] #define __noinline __attribute__((noinline))
[error]         ^
[error] /usr/include/sys/cdefs.h:358:9: note: previous definition is here
[error] #define __noinline      __attribute__ ((__noinline__))
[error]         ^
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:62:26: error: use of undeclared identifier 'ASM_JMPBUF_SIZE'
[error] typedef void *lh_jmp_buf[ASM_JMPBUF_SIZE / sizeof(void *)];
[error]                          ^
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:303:43: error: use of undeclared identifier 'ASM_JMPBUF_SIZE'
[error]     memcpy(return_buf, continuation->buf, ASM_JMPBUF_SIZE);
[error]                                           ^
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:323:5: error: use of undeclared identifier 'JMPBUF_STACK_POINTER_OFFSET'
[error]     jmpbuf_fix(return_buf);
[error]     ^
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:322:38: note: expanded from macro 'jmpbuf_fix'
[error] #define jmpbuf_fix(buf) fix_addr(buf[JMPBUF_STACK_POINTER_OFFSET])
[error]                                      ^
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:323:5: error: use of undeclared identifier 'JMPBUF_STACK_POINTER_OFFSET'
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:322:38: note: expanded from macro 'jmpbuf_fix'
[error] #define jmpbuf_fix(buf) fix_addr(buf[JMPBUF_STACK_POINTER_OFFSET])
[error]                                      ^
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:333:9: error: use of undeclared identifier 'JMPBUF_STACK_POINTER_OFFSET'
[error]         jmpbuf_fix(nw->h->buf);
[error]         ^
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:322:38: note: expanded from macro 'jmpbuf_fix'
[error] #define jmpbuf_fix(buf) fix_addr(buf[JMPBUF_STACK_POINTER_OFFSET])
[error]                                      ^
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:333:9: error: use of undeclared identifier 'JMPBUF_STACK_POINTER_OFFSET'
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:322:38: note: expanded from macro 'jmpbuf_fix'
[error] #define jmpbuf_fix(buf) fix_addr(buf[JMPBUF_STACK_POINTER_OFFSET])
[error]                                      ^
[error] 1 warning and 6 errors generated.
[info] Compiling to native code (10661 ms)
[info] Total (44294 ms)
[warn] In the last 937 seconds, 12.866 (1.4%) were spent in GC. [Heap: 1.41GB free of 2.02GB, max 12.00GB] Consider increasing the JVM heap using `-Xmx` or try a different collector, e.g. `-XX:+UseG1GC`, for better performance.
[error] Failed to compile /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/gc/shared/Safepoint.c
[error] (sandbox3 / Compile / nativeLink) Failed to compile /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/gc/shared/Safepoint.c
[error] Total time: 941 s (15:41), completed Oct 5, 2023 9:34:32 PM
```

--- a/docs/user/setup.rst
+++ b/docs/user/setup.rst
@@ -107,11 +107,6 @@ installation of FreeBSD.
 *Note 3:* Using the boehm GC with multi-threaded binaries doesn't work
 out-of-the-box yet.
 
-*Note 4:* A number of tests, primarily unit-tests, are known to fail
-or not terminate on FreeBSD. It is believed that the code-under-test
-is correct and that the defect is in the test itself.
-Work is underway to fix these tests.
-
 **Nix/NixOS**
 
 .. code-block:: shell

--- a/javalib/src/main/scala/java/net/Inet6Address.scala
+++ b/javalib/src/main/scala/java/net/Inet6Address.scala
@@ -31,8 +31,9 @@ final class Inet6Address private (
    */
 
   override def equals(that: Any): Boolean = that match {
-    case that: Inet6Address => this.hashCode() == that.hashCode()
-    case _                  => false
+    case that: Inet6Address =>
+      (that != null) && (this.hashCode() == that.hashCode())
+    case _ => false
   }
 
   def getScopedInterface(): NetworkInterface = nif
@@ -58,13 +59,21 @@ final class Inet6Address private (
   override def isLinkLocalAddress(): Boolean =
     (ipAddress(0) == -2) && ((ipAddress(1) & 255) >>> 6) == 2
 
-  override def isAnyLocalAddress(): Boolean = ipAddress.forall(_ == 0)
+  // avoid cost of functional style forall().
+  private def sumByteRange(bytes: Array[Byte], start: Int, end: Int): Int = {
+    // "end" is Java style exclusive, i.e. one past active range.
+    var count = 0
+    for (j <- start until end)
+      count += bytes(j)
+    count
+  }
+
+  override def isAnyLocalAddress(): Boolean =
+    sumByteRange(ipAddress, 0, 16) == 0
 
   override def isLoopbackAddress(): Boolean = {
-    if (ipAddress(15) != 1)
-      return false
-
-    ipAddress.dropRight(1).forall(_ == 0)
+    if ((ipAddress(0) != 0) || (ipAddress(15) != 1)) false
+    else sumByteRange(ipAddress, 2, 15) == 0
   }
 
   override def isMCGlobal(): Boolean =
@@ -87,16 +96,18 @@ final class Inet6Address private (
   override def isSiteLocalAddress(): Boolean =
     (ipAddress(0) == -2) && ((ipAddress(1) & 255) >>> 6) == 3
 
-  def isIPv4CompatibleAddress(): Boolean = ipAddress.take(12).forall(_ == 0)
+  def isIPv4CompatibleAddress(): Boolean =
+    sumByteRange(ipAddress, 0, 12) == 0
 
   private def formatScopeId(): String = {
     if (nif != null)
       nif.getDisplayName()
-    else if (useScopeId) {
+    else if (!useScopeId) ""
+    else {
       val netIf = NetworkInterface.getByIndex(scopeId)
       if (netIf == null) String.valueOf(scopeId)
       else netIf.getDisplayName()
-    } else ""
+    }
   }
 
 }

--- a/javalib/src/main/scala/java/net/InetAddress.scala
+++ b/javalib/src/main/scala/java/net/InetAddress.scala
@@ -743,7 +743,7 @@ object InetAddress {
          *   to creating an InetAddress using the hostname and the IPv4
          *   loopback address 127.0.0.1.  Be robust and do the same here.
          */
-        case e: UnknownHostException => SocketHelpers.loopbackIPv4()
+        case e: UnknownHostException => SocketHelpers.loopbackIPv4
       }
     }
   }

--- a/javalib/src/main/scala/java/net/NetworkInterface.scala
+++ b/javalib/src/main/scala/java/net/NetworkInterface.scala
@@ -78,8 +78,10 @@ import macOsIfDl._
 class NetworkInterface private (ifName: String) {
 
   override def equals(that: Any): Boolean = that match {
-    case that: NetworkInterface => this.hashCode() == that.hashCode()
-    case _                      => false
+    case that: NetworkInterface =>
+      if (that == null) false
+      else this.hashCode() == that.hashCode()
+    case _ => false
   }
 
   def getDisplayName(): String = getName()
@@ -256,7 +258,42 @@ object NetworkInterface {
     }
   }
 
-  private def createInetAddress(
+  /* Return an InetAddress created with regard for operating system
+   * practices.
+   *
+   * Linux and macOS always provides the scopeId/NetworkInterfaceId for
+   * InetAddresses used in the InterfaceAddressses created by NetworkInterface.
+   *
+   * FreeBSD provides that value only for IPv6 addresses which are not Global
+   * and not "special" (loopback, "any" a.k.a wildcard). The ipv4compatible
+   * (as distinct from the IPv4 mapped IPv6) address has not been verified
+   * since it is one step above deprecated and almost never used in the wild.
+   */
+  private def createOsConditionedInetAddress(
+      sa: Ptr[sockaddr],
+      ifIndex: Int
+  ): InetAddress = {
+    val af = sa.sa_family.toInt
+
+    val bytes = sockaddrToByteArray(sa) // will throw if != AF_INET or AF_INET6
+
+    if (af == AF_INET) {
+      InetAddress.getByAddress(bytes)
+    } else {
+      val provideScopeId =
+        if (!LinktimeInfo.isFreeBSD) true
+        else {
+          (((bytes(0) & 0xff) >>> 5) > 1) // top (leftmost) 2 bits are 00
+        }
+
+      if (!provideScopeId)
+        Inet6Address(bytes, "")
+      else
+        Inet6Address.getByAddress("", bytes, ifIndex)
+    }
+  }
+
+  private def createInetAddressOption(
       ifa: Ptr[ifaddrs],
       ifIndex: Int
   ): Option[InetAddress] = {
@@ -264,17 +301,10 @@ object NetworkInterface {
     val af = sa.sa_family.toInt
 
     if (!((af == AF_INET) || (af == AF_INET6))) None
-    else {
-      val bytes = sockaddrToByteArray(sa)
-      if (af == AF_INET) {
-        Some(InetAddress.getByAddress(bytes))
-      } else {
-        Some(Inet6Address.getByAddress("", bytes, ifIndex))
-      }
-    }
+    else Some(createOsConditionedInetAddress(sa, ifIndex))
   }
 
-  private def createInterfaceAddress(
+  private def createInterfaceAddressOption(
       ifa: Ptr[ifaddrs],
       ifIndex: Int
   ): Option[InterfaceAddress] = {
@@ -301,12 +331,7 @@ object NetworkInterface {
     if (!((af == AF_INET) || (af == AF_INET6))) {
       None // Silently skip AF_PACKET (17) and such.
     } else {
-      val bytes = sockaddrToByteArray(sa)
-      val inetAddress = if (af == AF_INET) {
-        InetAddress.getByAddress(bytes)
-      } else {
-        Inet6Address.getByAddress("", bytes, ifIndex)
-      }
+      val inetAddress = createOsConditionedInetAddress(sa, ifIndex)
 
       val broadcastAddress: Option[Array[Byte]] =
         if (sa.sa_family.toInt == AF_INET6) None
@@ -724,7 +749,7 @@ object NetworkInterface {
     ): Unit = {
       if (ifa != null) {
         if (strcmp(ifNameC, ifa.ifa_name) == 0) {
-          createInetAddress(ifa, ifIndex).map(ia => addOne(ia))
+          createInetAddressOption(ifa, ifIndex).map(ia => addOne(ia))
         }
         accumulateInetAddresses(
           ifNameC,
@@ -766,7 +791,7 @@ object NetworkInterface {
     ): Unit = {
       if (ifa != null) {
         if (strcmp(ifNameC, ifa.ifa_name) == 0) {
-          createInterfaceAddress(ifa, ifIndex).map(ia => addOne(ia))
+          createInterfaceAddressOption(ifa, ifIndex).map(ia => addOne(ia))
         }
         accumulateInterfaceAddresses(
           ifNameC,

--- a/javalib/src/main/scala/java/net/SocketHelpers.scala
+++ b/javalib/src/main/scala/java/net/SocketHelpers.scala
@@ -323,20 +323,22 @@ object SocketHelpers {
     new InetSocketAddress(addr, port)
   }
 
-  // Create copies of loopback & wildcard, so that originals never get changed
+  /* InetAddress() & Inet6Address() make defensive copies of the Array[Byte].
+   * As a result, these originals can never get changed.
+   */
 
   // ScalaJVM shows loopbacks with null host, wildcards with numeric host.
-  private[net] def loopbackIPv4(): InetAddress =
+  private[net] lazy val loopbackIPv4: InetAddress =
     InetAddress.getByAddress(Array[Byte](127, 0, 0, 1))
 
-  private[net] def loopbackIPv6(): InetAddress = InetAddress.getByAddress(
+  private[net] lazy val loopbackIPv6: InetAddress = InetAddress.getByAddress(
     Array[Byte](0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1)
   )
 
-  private def wildcardIPv4(): InetAddress =
+  private lazy val wildcardIPv4: InetAddress =
     InetAddress.getByAddress("0.0.0.0", Array[Byte](0, 0, 0, 0))
 
-  private def wildcardIPv6(): InetAddress = InetAddress.getByAddress(
+  private lazy val wildcardIPv6: InetAddress = InetAddress.getByAddress(
     "0:0:0:0:0:0:0:0",
     Array[Byte](0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)
   )
@@ -360,8 +362,8 @@ object SocketHelpers {
   }
 
   private[net] def getLoopbackAddress(): InetAddress = {
-    if (useLoopbackIPv6) loopbackIPv6()
-    else loopbackIPv4()
+    if (useLoopbackIPv6) loopbackIPv6
+    else loopbackIPv4
   }
 
   private lazy val useWildcardIPv6: Boolean = {
@@ -373,8 +375,8 @@ object SocketHelpers {
   }
 
   private[net] def getWildcardAddress(): InetAddress = {
-    if (useWildcardIPv6) wildcardIPv6()
-    else wildcardIPv4()
+    if (useWildcardIPv6) wildcardIPv6
+    else wildcardIPv4
   }
 
   /* Return the wildcard address corresponding directly to the IP stack in use.
@@ -387,9 +389,9 @@ object SocketHelpers {
    */
 
   private[net] def getWildcardAddressForBind(): InetAddress = {
-    if (LinktimeInfo.isFreeBSD) wildcardIPv4()
-    else if (useIPv4Stack) wildcardIPv4()
-    else wildcardIPv6()
+    if (LinktimeInfo.isFreeBSD) wildcardIPv4
+    else if (useIPv4Stack) wildcardIPv4
+    else wildcardIPv6
   }
 
   private[net] def fetchFdLocalAddress(osFd: Int): InetAddress = {

--- a/scripted-tests/run/build-library-dynamic/src/main/c/testlib.cpp
+++ b/scripted-tests/run/build-library-dynamic/src/main/c/testlib.cpp
@@ -42,17 +42,18 @@ int main() {
     }
     assert(exceptionCaught);
 
-#ifndef __APPLE__
-    // For some unknown reason on macOS our exception wrapper is not being
-    // caught. It works fine on Linux and Windows however.
-    // It's still possible to catch std::exception though
+#if !defined(__APPLE__) && !defined(__FreeBSD__)
+    // For some unknown reason on macOS or FreeBSD our exception wrapper is
+    // not being caught. It works fine on Linux and Windows however.
+    // It's still possible to catch std::exception though.
+
     exceptionCaught = false;
     try {
         fail();
     } catch (const scalanative::ExceptionWrapper &e) {
         exceptionCaught = true;
     }
-    assert(exceptionCaught);
+    assert(("exceptionCaught", exceptionCaught));
 #endif
 
     return 0;

--- a/unit-tests/native/src/test/scala/org/scalanative/testsuite/posixlib/LocaleTest.scala
+++ b/unit-tests/native/src/test/scala/org/scalanative/testsuite/posixlib/LocaleTest.scala
@@ -5,7 +5,7 @@ import org.junit.Assert._
 import org.junit.Assume._
 import org.junit.{BeforeClass, AfterClass}
 
-import scala.scalanative.meta.LinktimeInfo.{isLinux, isWindows}
+import scala.scalanative.meta.LinktimeInfo
 
 import scala.scalanative.unsafe._
 import scala.scalanative.unsigned._
@@ -23,10 +23,10 @@ object LocaleTest {
   def beforeClass(): Unit = {
     assumeTrue(
       "locale.scala is not implemented on Windows",
-      !isWindows
+      !LinktimeInfo.isWindows
     )
 
-    if (!isWindows) {
+    if (!LinktimeInfo.isWindows) {
       val entryLocale = setlocale(LC_ALL, null)
       assertNotNull(
         "setlocale() could not determine locale at start of test.",
@@ -53,7 +53,7 @@ object LocaleTest {
 
   @AfterClass
   def afterClass(): Unit = {
-    if (!isWindows) {
+    if (!LinktimeInfo.isWindows) {
       savedLocale.map { sl =>
         errno = 0
         // restore Locale as recorded on entry
@@ -82,7 +82,7 @@ class LocaleTest {
       savedLocale.isDefined
     )
 
-    if (!isWindows) {
+    if (!LinktimeInfo.isWindows) {
       val currentLconv = localeconv() // documented as always succeeds.
 
       assertEquals(
@@ -97,12 +97,21 @@ class LocaleTest {
         fromCString(currentLconv.thousands_sep)
       )
 
-      // Expect three byte-integers 3, 3, 0, meaning infinite group-by-three
-      assertEquals(
-        "US grouping",
-        "\u0003\u0003",
-        fromCString(currentLconv.grouping)
-      )
+      /* Skip grouping testing on FreeBSD. There is some long standing
+       * discussion that FreeBSD does not use POSIX compliant values.
+       * Do not test for an exact value that is known to be buggy.
+       * The is to reduce them maintenance headache & cost if that
+       * bug ever gets fixed.
+       */
+
+      if (!LinktimeInfo.isFreeBSD) {
+        // Expect three byte-integers 3, 3, 0  meaning infinite group-by-three
+        assertEquals(
+          "US grouping",
+          "\u0003\u0003",
+          fromCString(currentLconv.grouping)
+        )
+      }
 
       assertEquals(
         "US int_curr_symbol",
@@ -128,12 +137,16 @@ class LocaleTest {
         fromCString(currentLconv.mon_thousands_sep)
       )
 
-      // Expect three byte-integers 3, 3, 0, meaning infinite group-by-3
-      assertEquals(
-        "US mon_grouping",
-        "\u0003\u0003",
-        fromCString(currentLconv.mon_grouping)
-      )
+      // See "skip "FreeBSD"" comment before US grouping check above.
+
+      if (!LinktimeInfo.isFreeBSD) {
+        // Expect three byte-integers 3, 3, 0, meaning infinite group-by-3
+        assertEquals(
+          "US mon_grouping",
+          "\u0003\u0003",
+          fromCString(currentLconv.mon_grouping)
+        )
+      }
 
       assertEquals(
         "US positive_sign",
@@ -167,7 +180,7 @@ class LocaleTest {
 
       assertEquals("US int_n_cs_precedes", 1, currentLconv.int_n_cs_precedes)
 
-      if (isLinux) {
+      if (LinktimeInfo.isLinux) {
         assertEquals(
           "US int_p_sep_by_space",
           1,

--- a/unit-tests/native/src/test/scala/org/scalanative/testsuite/posixlib/MonetaryTest.scala
+++ b/unit-tests/native/src/test/scala/org/scalanative/testsuite/posixlib/MonetaryTest.scala
@@ -77,6 +77,8 @@ class MonetaryTest {
 
         val expected = if (Platform.isLinux()) {
           "[ $**1234.57] [ USD **1,234.57]"
+        } else if (Platform.isFreeBSD()) {
+          "[ **1234.57 ] [ **1234.57 ]"
         } else {
           "[ $**1234.57] [ USD**1,234.57]"
         }

--- a/unit-tests/native/src/test/scala/org/scalanative/testsuite/posixlib/sys/ResourceTest.scala
+++ b/unit-tests/native/src/test/scala/org/scalanative/testsuite/posixlib/sys/ResourceTest.scala
@@ -51,9 +51,9 @@ class ResourceTest {
 
     // Most operating systems will return EINVAL. Handle corner cases here.
     if (errno != EINVAL) {
-      if (!Platform.isLinux()) {
+      if (!(Platform.isLinux() || Platform.isFreeBSD())) {
         assertEquals("unexpected errno", EINVAL, errno)
-      } else if (errno != 0) { // Linux
+      } else if (errno != 0) { // Linux, FreeBSD
         // A pid of UInt.MaxValue is highly unlikely but, by one reading,
         // possible. If it exists and is found, it should not cause this test
         // to fail, just to be specious (have false look of genuineness).

--- a/unit-tests/native/src/test/scala/org/scalanative/testsuite/posixlib/sys/StatTest.scala
+++ b/unit-tests/native/src/test/scala/org/scalanative/testsuite/posixlib/sys/StatTest.scala
@@ -1,12 +1,12 @@
 package org.scalanative.testsuite.posixlib
+package sys
 
 import org.junit.Test
 import org.junit.Assert._
 import org.junit.Assume._
 import org.junit.BeforeClass
 
-// import scala.scalanative.meta.LinktimeInfo.{isLinux, isWindows}
-import scala.scalanative.meta.LinktimeInfo.isWindows
+import scala.scalanative.meta.LinktimeInfo
 
 import java.nio.file.{Files, Path}
 
@@ -29,7 +29,7 @@ object StatTest {
 
   @BeforeClass
   def beforeClass(): Unit = {
-    if (!isWindows) {
+    if (!LinktimeInfo.isWindows) {
       val orgDir = Files.createTempDirectory("scala-native-testsuite")
       val posixlibDir = orgDir.resolve("posixlib")
       workDirString = Files
@@ -42,7 +42,7 @@ object StatTest {
 class StatTest {
   import StatTest.workDirString
 
-  @Test def fileStatTest(): Unit = if (!isWindows) {
+  @Test def fileStatTest(): Unit = if (!LinktimeInfo.isWindows) {
     Zone { implicit z =>
       import scala.scalanative.posix.sys.statOps.statOps
 
@@ -82,9 +82,14 @@ class StatTest {
         statFromPath.st_rdev,
         statFromFd.st_rdev
       )
+
+      val expectedRdev =
+        if (!LinktimeInfo.isFreeBSD) 0.toUSize // Linux, macOS
+        else ULong.MaxValue.toUSize
+
       assertEquals(
-        "st_rdev must be 0 for regular file",
-        0.toUSize,
+        s"st_rdev must be ${expectedRdev} for regular file",
+        expectedRdev,
         statFromPath.st_rdev
       )
       assertEquals(
@@ -102,10 +107,25 @@ class StatTest {
         statFromPath.st_gid,
         statFromFd.st_gid
       )
+
       assertEquals("tmpfile must be empty", 0, statFromPath.st_size)
       assertEquals("tmpfile must be empty", 0, statFromFd.st_size)
-      assertEquals("tmpfile must not have blksize", 0, statFromPath.st_blocks)
-      assertEquals("tmpfile must not have blksize", 0, statFromFd.st_blocks)
+
+      val expectedBlksize =
+        if (!LinktimeInfo.isFreeBSD) 0 // Linux, macOS
+        else 1
+
+      assertEquals(
+        "unexpected statFromPath.blksize",
+        expectedBlksize,
+        statFromPath.st_blocks
+      )
+      assertEquals(
+        "unexpected statFromFd.blksize",
+        expectedBlksize,
+        statFromFd.st_blocks
+      )
+
       assertEquals(
         "st_atime from path and from fd must be the same",
         statFromPath.st_atime,
@@ -232,11 +252,13 @@ class StatTest {
         1,
         stat.S_ISDIR(dirStatFromPath.st_mode)
       )
+
       assertEquals(
-        "st_rdev must be 0 for dir",
-        0.toUSize,
+        s"st_rdev must be ${expectedRdev} for dir file",
+        expectedRdev,
         dirStatFromPath.st_rdev
       )
+
       assert(
         dirStatFromPath.st_nlink.toInt >= 2
       )

--- a/unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/net/InterfaceAddressTest.scala
+++ b/unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/net/InterfaceAddressTest.scala
@@ -28,20 +28,26 @@ class InterfaceAddressTest {
    * They also assume, perhaps unwisely, that the loopback address has index 1.
    */
 
-  val localhostIf =
+  val loopbackIfName =
     if (Platform.isLinux) "lo"
     else "lo0"
 
+  val loopbackIfIndex =
+    if (Platform.isFreeBSD) 2
+    else 1
+
   val osIPv6PrefixLength =
-    if (Platform.isMacOs) 64
+    if ((Platform.isMacOs) || (Platform.isFreeBSD)) 64
     else 128
 
   val osIPv6LoopbackSuffix =
-    s":0:0:0:0:0:0:1%${localhostIf}"
+    s":0:0:0:0:0:0:1%${loopbackIfName}"
 
   val osIPv6LoopbackAddress =
-    if (Platform.isMacOs) s"fe80${osIPv6LoopbackSuffix}"
-    else s"0${osIPv6LoopbackSuffix}"
+    if ((Platform.isMacOs) || (Platform.isFreeBSD))
+      s"fe80${osIPv6LoopbackSuffix}"
+    else
+      s"0${osIPv6LoopbackSuffix}"
 
   /* Test equals() but there is no good, simple way to test corresponding
    * hashCode(). The contents of the components used in the hash vary by
@@ -52,7 +58,7 @@ class InterfaceAddressTest {
   @Test def testEquals(): Unit = {
     assumeFalse("Not implemented in Windows", Platform.isWindows)
 
-    val netIf = NetworkInterface.getByIndex(1) // loopback
+    val netIf = NetworkInterface.getByIndex(loopbackIfIndex)
     assertNotNull(netIf)
 
     val ifAddresses = netIf.getInterfaceAddresses()
@@ -70,7 +76,7 @@ class InterfaceAddressTest {
   @Test def testGetAddress(): Unit = {
     assumeFalse("Not implemented in Windows", Platform.isWindows)
 
-    val netIf = NetworkInterface.getByIndex(1) // loopback
+    val netIf = NetworkInterface.getByIndex(loopbackIfIndex)
     assertNotNull(netIf)
 
     val ifAddresses = netIf.getInterfaceAddresses()
@@ -85,7 +91,10 @@ class InterfaceAddressTest {
           if (!hostAddr.contains(":")) {
             "127.0.0.1"
           } else if (hostAddr.startsWith("0")) {
-            s"0:0:0:0:0:0:0:1%${localhostIf}"
+            val suffix =
+              if (Platform.isFreeBSD) ""
+              else s"%${loopbackIfName}"
+            s"0:0:0:0:0:0:0:1${suffix}"
           } else if (hostAddr.startsWith("f")) {
             s"${osIPv6LoopbackAddress}"
           } else "" // fail in a way that will print out ifAddrString
@@ -106,7 +115,7 @@ class InterfaceAddressTest {
   @Test def testGetNetworkPrefixLength(): Unit = {
     assumeFalse("Not implemented in Windows", Platform.isWindows)
 
-    val netIf = NetworkInterface.getByIndex(1) // loopback
+    val netIf = NetworkInterface.getByIndex(loopbackIfIndex)
     assertNotNull(netIf)
 
     val ifAddresses = netIf.getInterfaceAddresses()
@@ -139,7 +148,7 @@ class InterfaceAddressTest {
      *     InetAddress / prefix length [ broadcast address ]
      */
 
-    val netIf = NetworkInterface.getByIndex(1) // loopback
+    val netIf = NetworkInterface.getByIndex(loopbackIfIndex)
     assertNotNull(netIf)
 
     val ifAddresses = netIf.getInterfaceAddresses()
@@ -155,7 +164,10 @@ class InterfaceAddressTest {
           if (!ifAddrString.contains(":")) {
             "/127.0.0.1/8 [null]"
           } else if (ifAddrString.startsWith("/0")) {
-            s"/0:0:0:0:0:0:0:1%${localhostIf}/128 [null]"
+            val stem =
+              if (Platform.isFreeBSD) ""
+              else s"%${loopbackIfName}"
+            s"/0:0:0:0:0:0:0:1${stem}/128 [null]"
           } else if (ifAddrString.startsWith("/f")) {
             s"/${osIPv6LoopbackAddress}/${osIPv6PrefixLength} [null]"
           } else "" // fail in a way that will print out ifAddrString

--- a/unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/net/NetworkInterfaceTest.scala
+++ b/unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/net/NetworkInterfaceTest.scala
@@ -5,6 +5,7 @@ import java.net._
 import org.junit.Test
 import org.junit.Assert._
 import org.junit.Assume._
+import org.junit.BeforeClass
 
 import org.scalanative.testsuite.utils.AssertThrows.assertThrows
 import org.scalanative.testsuite.utils.Platform
@@ -20,23 +21,36 @@ import org.scalanative.testsuite.utils.Platform
  *       editing to reflect that local configuration.
  */
 
+object NetworkInterfaceTest {
+  @BeforeClass
+  def beforeClass(): Unit = {
+
+    assumeFalse("Not implemented in Windows", Platform.isWindows)
+  }
+}
+
 class NetworkInterfaceTest {
 
-  val localhostIf =
+  val loopbackIfName =
     if (Platform.isLinux) "lo"
     else "lo0"
 
+  val loopbackIfIndex =
+    if (Platform.isFreeBSD) 2
+    else 1
+
   val osIPv6LoopbackSuffix =
-    s":0:0:0:0:0:0:1%${localhostIf}"
+    s":0:0:0:0:0:0:1%${loopbackIfName}"
 
   val osIPv6LoopbackAddress =
-    if (Platform.isMacOs) s"fe80${osIPv6LoopbackSuffix}"
-    else s"0${osIPv6LoopbackSuffix}"
+    if ((Platform.isMacOs) || (Platform.isFreeBSD))
+      s"fe80${osIPv6LoopbackSuffix}"
+    else
+      s"0${osIPv6LoopbackSuffix}"
 
 // Test static (object) methods
 
   @Test def getByIndexMinusTwo(): Unit = {
-    assumeFalse("Not implemented in Windows", Platform.isWindows)
     assertThrows(
       "getByIndex(-2)",
       classOf[IllegalArgumentException],
@@ -45,29 +59,27 @@ class NetworkInterfaceTest {
   }
 
   @Test def getByIndexZero(): Unit = {
-    assumeFalse("Not implemented in Windows", Platform.isWindows)
     assertNull(NetworkInterface.getByIndex(0))
   }
 
   @Test def getByIndexOne(): Unit = {
-    assumeFalse("Not implemented in Windows", Platform.isWindows)
-    val netIf = NetworkInterface.getByIndex(1) // loopback
+    val netIf = NetworkInterface.getByIndex(1)
 
     assertNotNull("a1", netIf)
 
-    val sought = localhostIf
+    val sought =
+      if (Platform.isFreeBSD) "em0"
+      else loopbackIfName
     val ifName = netIf.getName()
     assertEquals("a2", sought, ifName)
   }
 
   @Test def getByIndexMaxValue(): Unit = {
-    assumeFalse("Not implemented in Windows", Platform.isWindows)
     val netIf = NetworkInterface.getByIndex(Integer.MAX_VALUE)
     assertNull("Unlikely interface found for MAX_VALUE index", netIf)
   }
 
   @Test def getByInetAddressNull(): Unit = {
-    assumeFalse("Not implemented in Windows", Platform.isWindows)
     assertThrows(
       "getByInetAddress(null)",
       classOf[NullPointerException],
@@ -76,34 +88,31 @@ class NetworkInterfaceTest {
   }
 
   @Test def getByInetAddressLoopbackIPv4(): Unit = {
-    assumeFalse("Not implemented in Windows", Platform.isWindows)
     val lba4 = InetAddress.getByName("127.0.0.1")
 
     val netIf = NetworkInterface.getByInetAddress(lba4)
 
     assertNotNull("a1", netIf)
 
-    val sought = localhostIf
+    val sought = loopbackIfName
     val ifName = netIf.getName()
     assertEquals("a1", sought, ifName)
   }
 
   @Test def getByInetAddressLoopbackIPv6(): Unit = {
-    assumeFalse("Not implemented in Windows", Platform.isWindows)
     val lba6 = InetAddress.getByName("::1")
 
     val netIf = NetworkInterface.getByInetAddress(lba6)
 
     // Do not fail on null. IPv6 might not be enabled on the system.
     if (netIf != null) {
-      val sought = localhostIf
+      val sought = loopbackIfName
       val ifName = netIf.getName()
       assertEquals("a1", sought, ifName)
     }
   }
 
   @Test def getByNameNull(): Unit = {
-    assumeFalse("Not implemented in Windows", Platform.isWindows)
     assertThrows(
       "getByName(null)",
       classOf[NullPointerException],
@@ -112,9 +121,7 @@ class NetworkInterfaceTest {
   }
 
   @Test def getByName(): Unit = {
-    assumeFalse("Not implemented in Windows", Platform.isWindows)
-
-    val sought = localhostIf
+    val sought = loopbackIfName
     val netIf = NetworkInterface.getByName(sought)
     assertNotNull(netIf)
 
@@ -124,9 +131,7 @@ class NetworkInterfaceTest {
   }
 
   @Test def testToString(): Unit = {
-    assumeFalse("Not implemented in Windows", Platform.isWindows)
-
-    val netIf = NetworkInterface.getByIndex(1) // loopback
+    val netIf = NetworkInterface.getByIndex(loopbackIfIndex)
     assertNotNull(netIf)
 
     val ifName = netIf.getName()
@@ -138,8 +143,6 @@ class NetworkInterfaceTest {
   }
 
   @Test def getNetworkInterfaces(): Unit = {
-    assumeFalse("Not implemented in Windows", Platform.isWindows)
-
     val netIfs = NetworkInterface.getNetworkInterfaces()
     assertNotNull(netIfs)
 
@@ -157,11 +160,9 @@ class NetworkInterfaceTest {
 // Test instance methods
 
   @Test def instanceGetIndex(): Unit = {
-    assumeFalse("Not implemented in Windows", Platform.isWindows)
-
-    val lbIf1 = NetworkInterface.getByName(localhostIf)
+    val lbIf1 = NetworkInterface.getByName(loopbackIfName)
     assertNotNull(lbIf1)
-    assertEquals(1, lbIf1.getIndex())
+    assertEquals(loopbackIfIndex, lbIf1.getIndex())
   }
 
   /*  @Test def instanceGetHardwareAddress(): Unit = {
@@ -171,9 +172,7 @@ class NetworkInterfaceTest {
    */
 
   @Test def instanceGetMTU(): Unit = {
-    assumeFalse("Not implemented in Windows", Platform.isWindows)
-
-    val lbIf = NetworkInterface.getByName(localhostIf)
+    val lbIf = NetworkInterface.getByName(loopbackIfName)
     assertNotNull(lbIf)
 
     val mtu = lbIf.getMTU()
@@ -184,47 +183,36 @@ class NetworkInterfaceTest {
   }
 
   @Test def instanceIsLoopback(): Unit = {
-    assumeFalse("Not implemented in Windows", Platform.isWindows)
-
-    val lbIf = NetworkInterface.getByName(localhostIf)
+    val lbIf = NetworkInterface.getByName(loopbackIfName)
     assertNotNull(lbIf)
     assertEquals("a1", true, lbIf.isLoopback())
   }
 
   @Test def instanceIsPoinToPoint(): Unit = {
-    assumeFalse("Not implemented in Windows", Platform.isWindows)
-
-    val lbIf = NetworkInterface.getByName(localhostIf)
+    val lbIf = NetworkInterface.getByName(loopbackIfName)
     assertNotNull(lbIf)
     assertEquals("a1", false, lbIf.isPointToPoint())
   }
 
   @Test def instanceIsUp(): Unit = {
-    assumeFalse("Not implemented in Windows", Platform.isWindows)
-
-    val lbIf = NetworkInterface.getByName(localhostIf)
+    val lbIf = NetworkInterface.getByName(loopbackIfName)
     assertNotNull(lbIf)
     assertEquals("a1", true, lbIf.isUp())
   }
 
   @Test def instanceSupportsMulticast(): Unit = {
-    assumeFalse("Not implemented in Windows", Platform.isWindows)
-
-    val lbIf = NetworkInterface.getByName(localhostIf)
+    val lbIf = NetworkInterface.getByName(loopbackIfName)
     assertNotNull(lbIf)
 
     val expected =
-      if (Platform.isMacOs) true
+      if ((Platform.isMacOs) || (Platform.isFreeBSD)) true
       else false // Linux
-    // else (FreeBSD?)
 
     assertEquals("a1", expected, lbIf.supportsMulticast())
   }
 
   @Test def instanceGetInetAddresses(): Unit = {
-    assumeFalse("Not implemented in Windows", Platform.isWindows)
-
-    val lbIf = NetworkInterface.getByName(localhostIf)
+    val lbIf = NetworkInterface.getByName(loopbackIfName)
     assertNotNull(lbIf)
 
     val iaEnumeration = lbIf.getInetAddresses()
@@ -239,7 +227,9 @@ class NetworkInterfaceTest {
         if (!hostAddr.contains(":")) {
           "127.0.0.1"
         } else if (hostAddr.startsWith("0")) {
-          s"0:0:0:0:0:0:0:1%${localhostIf}"
+          val stem = "0:0:0:0:0:0:0:1"
+          if (Platform.isFreeBSD) stem
+          else s"${stem}%${loopbackIfName}"
         } else if (hostAddr.startsWith("f")) {
           s"${osIPv6LoopbackAddress}"
         } else "" // fail in a way that will print out ifAddrString

